src/cirrus/lib/process_payload.py

```diff
@@ -277,14 +333,58 @@ def payload_ids(self) -> list[str]:
         """
         return [c["id"] for c in self.payloads]
 
-    def get_states(self) -> dict[str, StateEnum]:
+    def get_states_and_exec_arn(self) -> dict[str, tuple]:
         if self.state_items is None:
             items = [
                 self.statedb.dbitem_to_item(i)
                 for i in self.statedb.get_dbitems(self.payload_ids)
             ]
             self.state_items = items
-        return {c["payload_id"]: StateEnum(c["state"]) for c in self.state_items}
+        response = dict(self.get_process_attrs(self.state_items))
+        response.update(
+            {
+                p["id"]: (
+                    None,
+                    ProcessPayloads.gen_execution_arn(p["id"], p.process["workflow"]),
+                )
+                for p in self.payloads
+                if p["id"] not in response
+            },
+        )
+        return response
+
+    @staticmethod
+    def get_process_attrs(state_items):
+        for si in state_items:
+            payload_id = si["payload_id"]
+            state = StateEnum(si["state"]) if si["state"] is not None else None
+
+            # if workflow is found in CLAIMED state, then we want the last execution to
+            # determine where in the process of sfn execution the failure to proceeed to
+            # PROCESSING occurred.
+            if state == StateEnum.CLAIMED:
+                exec_arn = si["executions"][-1]
+            else:
+                exec_arn = ProcessPayloads.gen_execution_arn(
+                    payload_id,
+                    si["workflow"],
+                )
+
+            yield payload_id, (state, exec_arn)
+
+    @staticmethod
+    def gen_execution_arn(payload_id, workflow) -> str:
+        """
+        Generate an execution arn for the given payload_id, using the state_item info if
+        given.
+        """
+        # is this the right UUID, or do we want to have an idempotent one
+        # that includes the executions: {payload_id}/{state_item['executions']}?
```

Comment:
I like this idea. It makes a lot of sense, particularly given the use of uuid5 here is very specifically to enable deterministic uuids, which we don't really have with some (effectively) random timestamp.

Reply:
will do.